// eslint-plugin-jsdoc/dist/rules/validTypes.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _jsdoctypeparser = require("jsdoctypeparser");

var _iterateJsdoc = _interopRequireDefault(require("../iterateJsdoc"));

var _warnRemovedSettings = _interopRequireDefault(require("../warnRemovedSettings"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const asExpression = /as\s+/;

var _default = (0, _iterateJsdoc.default)(({
  jsdoc,
  report,
  utils,
  context
}) => {
  (0, _warnRemovedSettings.default)(context, 'valid-types');

  const _ref = context.options[0] || {},
        _ref$allowEmptyNamepa = _ref.allowEmptyNamepaths,
        allowEmptyNamepaths = _ref$allowEmptyNamepa === void 0 ? true : _ref$allowEmptyNamepa,
        _ref$checkSeesForName = _ref.checkSeesForNamepaths,
        checkSeesForNamepaths = _ref$checkSeesForName === void 0 ? false : _ref$checkSeesForName;

  if (!jsdoc.tags) {
    return;
  }

  jsdoc.tags.forEach(tag => {
    const validTypeParsing = function validTypeParsing(type, tagName) {
      try {
        (0, _jsdoctypeparser.parse)(type);
      } catch (err) {
        let error = err;

        if (tagName) {
          if (['memberof', 'memberof!'].includes(tagName)) {
            const endChar = type.slice(-1);

            if (['#', '.', '~'].includes(endChar)) {
              try {
                (0, _jsdoctypeparser.parse)(type.slice(0, -1));
                error = {};
              } catch (memberofError) {// Use the original error for including the whole type
              }
            }
          } else if (tagName === 'borrows') {
            const startChar = type.charAt();

            if (['#', '.', '~'].includes(startChar)) {
              try {
                (0, _jsdoctypeparser.parse)(type.slice(1));
                error = {};
              } catch (memberofError) {// Use the original error for including the whole type
              }
            }
          }
        }

        if (error.name === 'SyntaxError') {
          report(`Syntax error in type: ${type}`, null, tag);
          return false;
        }
      }

      return true;
    };

    if (tag.tag === 'borrows') {
      const thisNamepath = tag.description.replace(asExpression, '');

      if (!asExpression.test(tag.description) || !thisNamepath) {
        report(`@borrows must have an "as" expression. Found "${tag.description}"`, null, tag);
        return;
      }

      if (validTypeParsing(thisNamepath, 'borrows')) {
        const thatNamepath = tag.name;
        validTypeParsing(thatNamepath);
      }
    } else if (utils.isNamepathTag(tag.tag, checkSeesForNamepaths)) {
      if (utils.passesEmptyNamepathCheck(tag, allowEmptyNamepaths)) {
        return;
      }

      validTypeParsing(tag.name, tag.tag);
    } else if (tag.type && utils.isTagWithType(tag.tag)) {
      validTypeParsing(tag.type);
    }
  });
}, {
  iterateAllJsdocs: true,
  meta: {
    schema: [{
      additionalProperies: false,
      properties: {
        allowEmptyNamepaths: {
          default: true,
          type: 'boolean'
        },
        checkSeesForNamepaths: {
          default: false,
          type: 'boolean'
        }
      },
      type: 'object'
    }],
    type: 'suggestion'
  }
});

exports.default = _default;
module.exports = exports.default;
